src/coreclr/interpreter/compiler.cpp

```diff
@@ -7945,6 +7980,22 @@ void InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
 
                 if (m_compHnd->isValueClass(resolvedToken.hClass))
                 {
+                    CorInfoType asCorInfoType = m_compHnd->asCorInfoType(m_compHnd->getTypeForBox(resolvedToken.hClass));
+                    if (asCorInfoType == CORINFO_TYPE_FLOAT && m_pStackPointer[-1].GetStackType() == StackTypeR8)
+                    {
+                        EmitConv(m_pStackPointer - 1, StackTypeR4, INTOP_CONV_R4_R8);
+                    }
+                    else if (asCorInfoType == CORINFO_TYPE_DOUBLE && m_pStackPointer[-1].GetStackType() == StackTypeR4)
+                    {
+                        EmitConv(m_pStackPointer - 1, StackTypeR8, INTOP_CONV_R8_R4);
+                    }
+#ifdef TARGET_64BIT
+                    // nint and int32 can be used interchangeably. Add implicit conversions.
+                    else if ((asCorInfoType == CORINFO_TYPE_INT || asCorInfoType == CORINFO_TYPE_UINT) && m_pStackPointer[-1].GetStackType() == StackTypeI4)
```

Comment:
Shouldn't we use here the nint/nuint variants ? (CORINFO_TYPE_NATIVEINT and CORINFO_TYPE_NATIVEUNT). Also, should we use `INTOP_CONV_I8_U4` for the unsigned case ?